libs/api/domains/license-service/src/lib/client/driving-license-client/pkpass.client.ts

```diff
@@ -0,0 +1,445 @@
+  private readonly pkpassCacheTokenExpiryDelta: number
+  private readonly pkpassAuthRetries: number
+
+  constructor(
+    private config: Config,
+    private logger: Logger,
+    private cacheManager?: CacheManager | null,
+  ) {
+    this.pkpassApiKey = config.pkpass.apiKey
+    this.pkpassSecretKey = config.pkpass.secretKey
+    this.pkpassApiUrl = config.pkpass.apiUrl
+    this.pkpassCacheKey = config.pkpass.cacheKey
+    this.pkpassCacheTokenExpiryDelta =
+      strToPositiveNum(config.pkpass.cacheTokenExpiryDelta) ??
+      DEFAULT_CACHE_TOKEN_EXPIRY_DELTA_IN_MS
+    this.pkpassAuthRetries =
+      strToPositiveNum(config.pkpass.authRetries) ?? DEFAULT_AUTH_RETRIES
+
+    this.logger = logger
+    this.cacheManager = cacheManager
+  }
+
+  private parseTtlFromTokenExpiry(expiry?: string | null): number | null {
+    if (!expiry) {
+      return null
+    }
+
+    try {
+      const parsed: number =
+        Date.parse(expiry) - Date.now() - this.pkpassCacheTokenExpiryDelta
+
+      if (parsed > 0) {
+        return parsed
+      }
+    } catch (e) {
+      this.logger.verbose('unable to parse datetime from token', {
+        exception: e,
+        category: LOG_CATEGORY,
+      })
+    }
+
+    return null
+  }
+
+  private async getCachedGetPkPassToken(): Promise<string | null> {
+    if (!this.cacheManager) {
+      return null
+    }
+
+    const cached = await this.cacheManager.get(this.pkpassCacheKey)
+
+    if (cached && typeof cached === 'string') {
+      return cached
+    }
+
+    return null
+  }
+
+  private urlFetch(
+    payload: PkPassPayload,
+  ): (accessToken: string) => Promise<Response> {
+    return (accessToken: string): Promise<Response> => {
+      return fetch(`${this.pkpassApiUrl}/v2/driversLicense`, {
+        method: 'POST',
+        headers: {
+          apiKey: this.pkpassApiKey,
+          accessToken: `smart ${accessToken}`,
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(payload),
+      })
+    }
+  }
+
+  private async fetchRetried(
+    fetchToPerform: (token: string) => Promise<Response | null>,
+    retries = 0,
+  ): Promise<Response | null> {
+    let count = 0
+    const actualRetries = 0 <= retries && retries <= 3 ? retries : 0
+
+    if (retries !== actualRetries) {
+      this.logger.warn(
+        `service retry count (${retries}) outside of bounds [0,3], setting as 0`,
+        {
+          category: LOG_CATEGORY,
+        },
+      )
+    }
+
+    do {
+      // For retries, force a non-cached token
+      const token = await this.getPkPassToken(count > 0)
+
+      if (!token) {
+        throw new Error('null from getPkPassToken')
+      }
+
+      const res = await fetchToPerform(token)
+
+      if (res?.status !== 401) {
+        return res
+      }
+
+      count = count + 1
+      this.logger.info(`401 from service, retry number ${count}`, {
+        category: LOG_CATEGORY,
+      })
+    } while (count < actualRetries)
```

Comment:
I suspect it is because `actualRetries` is by default 1. Then this do while loop will only run once.
Because the loop runs only once, it will always use the token from the cache and never refresh it when it gets 401.

I am trying out changing the environment variable that controls actualRetries to more than 1. Then it should refresh the token when it is invalid.

Reply:
I works when I set `PKPASS_AUTH_RETRIES=3`